### PR TITLE
feat: Plugin input variables are not editable

### PR DESCRIPTION
**Pull request by dhananjay0601**
# Description

We are not able to edit the plugin input variables. When we want to edit any input variable , it is not allowing me to do the same.

Fixes https://github.com/devtron-labs/devtron/issues/3833, https://github.com/devtron-labs/devtron/issues/3424
          

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [X] editing the plugin input variables field.

# Checklist:

* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [X] I have performed a self-review of my own code
* [X] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 1]
```diff
@@ -198,14 +198,16 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
...
+     
     const ValueContainer = (props) => {
         let value = props.getValue()[0]?.label
+        const flag = !props.selectProps.menuIsOpen
         return (
             <components.ValueContainer {...props}>
                 <>
-                    {!props.selectProps.menuIsOpen &&
-                        (value ? `${value}` : <span className="cn-5">Select source or input value</span>)}
+                    {value ? `${value}` : flag && (
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:208]: {value ? `${value}` : flag  here we can directly use value without interpolation

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 2]
```diff
@@ -200,14 +200,15 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
             )
         }
     }
-
     const ValueContainer = (props) => {
         let value = props.getValue()[0]?.label
+        const flag = !props.selectProps.menuIsOpen
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:205]: Can use direct value of flag

File: src/setupProxy.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ module.exports = function (app) {
     app.use(
         '/orchestrator',
         createProxyMiddleware({
```
Comment (shivani170) [inline src/setupProxy.js:6]: revert please

**Review by shivani170 — DISMISSED** [review]

**Comment by shivani170** [comment]
Please fix semantic error

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 4]
```diff
@@ -1,20 +1,32 @@
...
 import { OptionType } from '../app/types'
 import { pipelineContext } from '../workflowEditor/workflowEditor'
 import { excludeVariables } from './Constants'
+import { Label } from 'recharts'
+import { MenuList } from 'react-select/dist/declarations/src/components/Menu'
+import AboutTagEditModal from '../app/details/AboutTagEditModal'
+import { importComponentFromFELibrary } from '../common/helpers/Helpers'
+import {TagLabelValueSelector} from '@devtron-labs/devtron-fe-common-lib'
+import { InputPluginSelection } from './InputPluginSelect'
+// import { Select } from '../common'
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:24]: Remove commented code please

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 5]
```diff
@@ -1,20 +1,32 @@
...
+
 
 function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariableIndex: number }) {
+    const refVar = useRef(null)
+    const dependentRef = useRef(null)
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:29]: `dependentRef` is not required

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 6]
```diff
@@ -30,17 +42,18 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
     const [selectedOutputVariable, setSelectedOutputVariable] = useState<OptionType>({
         label: '',
         value: '',
+        format: '',
     })
+   
+    const [inputVariableOptions, setInputVariableOptions] = useState<{
+        label: string,
+        options: any[]
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:50]: Please provide the type

File: src/components/ConfigMapSecret/Types.ts [hunk 7]
```diff
@@ -14,6 +15,24 @@ export interface KeyValueInputInterface {
     valueType?: string
 }
 
+export interface SuggestedTagOptionType extends OptionType {
+    label: string
+    options: any[]
```
Comment (shivani170) [inline src/components/ConfigMapSecret/Types.ts:20]: Please provide type

File: src/components/app/types.ts [hunk 8]
```diff
@@ -11,6 +11,7 @@ export interface AddNewAppProps extends RouteComponentProps<{}> {
 export interface OptionType {
     label: string
     value: string
+    format?: string
```
Comment (shivani170) [inline src/components/app/types.ts:14]: extend `format` type instead

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 9]
```diff
@@ -253,43 +270,17 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
             event.target.blur()
         }
     }
-
+    console.log(inputVariableOptions)
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:273]: Remove console

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 10]
```diff
@@ -253,43 +270,17 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
...
-            onKeyDown={handleKeyDown}
-            menuPlacement="auto"
+        <InputPluginSelection
+            placeholder='Select source or input value'
+            selectedOutputVariable={selectedOutputVariable}
+            setTagData={handleOutputVariableSelector}
+            tagData={selectedOutputVariable}
+            refVar={refVar}
+            dependentRef={dependentRef}
+            // tagOptions={inputVariableOptions}
```
Comment (shivani170) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:282]: remove commented code

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 11]
```diff
@@ -30,17 +29,14 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
     const [selectedOutputVariable, setSelectedOutputVariable] = useState<OptionType>({
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:29]: Please fix the type

File: src/components/CIPipelineN/CustomInputVariableSelect.tsx [hunk 12]
```diff
@@ -138,7 +136,7 @@ function CustomInputVariableSelect({ selectedVariableIndex }: { selectedVariable
         } else {
             _variableDetail = {
                 variableType: RefVariableType.NEW,
-                value: selectedValue.label,
+                value: selectedValue?.label,
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/CustomInputVariableSelect.tsx:139]: please remove this unnecessary `?` check

File: src/components/ConfigMapSecret/Types.ts [hunk 13]
```diff
@@ -14,6 +15,33 @@ export interface KeyValueInputInterface {
...
+
+export interface InputPluginSelectionType { 
+    selectedOutputVariable: OptionType
+    variableOptions?: SuggestedTagOptionType[]
+    variableData?: OptionType
+    setVariableData?: (tagData:OptionType ) => void
+    refVar?: React.MutableRefObject<HTMLTextAreaElement>
+    noBackDrop?: boolean
+    placeholder: string
+    dependentRef?: React.MutableRefObject<HTMLTextAreaElement>
```
Comment (vivek-devtron) [inline src/components/ConfigMapSecret/Types.ts:40]: is this required here?

File: src/components/app/types.ts (hunk 8 above)
Comment (vivek-devtron) [reply src/components/app/types.ts:14]: We should not add the format in generic type. we should extend this for our specific use case

File: src/components/CIPipelineN/InputPluginSelect.tsx [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+                    value={selectedValue}
+                    onChange={handleInputChange}
+                    onBlur={handleOnBlur}
+                    placeholder={placeholder}
+                    refVar={refVar}
+                    tabIndex={selectedVariableIndex}
+                    handleKeyDown={handleOnKeyDown}
+                />
+            </PopupMenu.Button>
+            {selectedValue ? (
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/InputPluginSelect.tsx:175]: this can be written as {selectedValue && (

**Review by vivek-devtron — DISMISSED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**